### ISSUE TITLE
ECDSA should have a (optional?) hash parameter

### ISSUE THREAD

**Issue by HLRichardson-Git** [HLRichardson-Git/Gestalt#29]
I will pull this directly from the README as this is the intended way that ECDSA signing and verifying works:
```cpp
#include <gestalt/ecdsa.h>
#include <gestalt/sha2.h>
#include <iostream>

int main() {
    std::string privateKey = "0xC9AFA9D845BA75166B5C215767B1D6934E50C3DB36E89B127B8A622B120F6721";
    ECDSA ecdsa(StandardCurve::P256, privateKey);

    std::string message = "Hello, Gestalt!";
    std::string messageHash = hashSHA256(message);

    Signature signature = ecdsa.signMessage(messageHash);
    bool signatureStatus = ecdsa.verifySignature(messageHash, signature);

    if (signatureStatus) std::cout << "Valid!" << std::endl;

    return 0;
}
```
Currently, the `signMessage` and `verifySignature` functions expect the inputs to already be the hash of the message they want to sign or verify. This is not consistent with how we have been implementing RSA which expects the user to give the wanted hash algorithm for the message as a parameter. So in RSA they can give the message ciphertext and it will be hashed according to the default/ their given hash function they want to use. 

I propose we do the same here for ECDSA. In both `signMessage` and `verifySignature` another parameter should be needed which is an enum for the user to specify what hashing algorithm they want the message to be hashed with. Another thing I need to think of is if this additional hash algorithm parameter should be optional, so if one is not chosen it defaults to SHA256 for example. I haven't decided yet.

**Comment by mirzaim** [comment]
Hi Hunter,

I’d love to take on this feature! Could you assign it to me?

I’ve gone through the code and put together a preliminary version that updates the ECDSA signMessage and verifySignature functions to include the option for specifying a hash algorithm. Here’s the initial implementation: [link to commit](https://github.com/mirzaim/Gestalt/commit/6d7bc1c7ab8d1fde1a992e9be092dba09d7287d6).

Does this approach align with what you’re looking for? If so, I’d be glad to keep refining it and run any needed tests.

Thanks

**Comment by HLRichardson-Git** [comment]
Hello Mirzaim!

First off I would like to thank you for taking an interest in contributing to Gestalt. Funnily I actually just started on this issue, but you took a similar approach I was going about. I would be happy to assign this to you to complete.

### Details

I notice that your commit does not include the latest change I made yesterday in this [commit](https://github.com/HLRichardson-Git/Gestalt/commit/7960e8e5ab2d966bd6e44c10f781b8c8d175a2bf). It adds a `const ECDSAPublicKey& peerPublicKey` to the `verifySignature` function. Could you pull this commit?
 
I quite like how you did `getHashFunction` which is way more elegant then the way I did it for RSA.

There is one thing to note, just based on how I remember I did the unit tests for ECDSA. I recall didn't want the unit tests to use Gestalts SHA in its test to hash the message to be signed as I thought it would be confusing if we messed up the SHA implementation and we saw ECDSA operations failing as well. To that end the test vectors we have in `tests/ecc/vectors/vectors_ecdsa.h` have the hash of a message hard coded. So with this addition, these test vectors would fail as it would hash the hash if that makes sense.

We could either do:
1. Keep the `Signature signMessage(const std::string& messageHash);` and `verifySignature(const std::string& messageHash, const ECDSAPublicKey& peerPublicKey, const Signature signature);` in addition to your overloaded functions. These would be used as they are currently, expecting that the input is already the hash of a message to be signed.
2. The file with the test vectors: `tests/ecc/vectors/vectors_ecdsa.h` has the source I got the vectors from [here](https://www.rfc-editor.org/rfc/rfc6979#page-33). So we could change the `std::string msg;` of each vector to the message which I think is `message = "sample"`

I think my preference is the first as it just gives more options to users that could be useful in some edge cases. But I can be swayed easily either way honestly.

### Testing

Yes, I would like for you to confirm the unit tests we currently have are still passing. Building Gestalt should currently use these two flags: `GESTALT_BUILD_TESTS` and `GESTALT_FETCH_GOOGLETEST` and should compile a binary `tests.exe` which when ran should run all the unit tests.

If you cannot get it working, I'd be happy to hop in a call and try to troubleshoot. Additionally, when we have a good commit that may close this issue I will pull it and run the tests myself anyways.

Let me know if you have any questions, I would be happy to help and thanks again.

**Comment by mirzaim** [comment]
Hi Hunter,

I wanted to share my latest approach for this feature: [Commit Link](https://github.com/mirzaim/Gestalt/commit/bff5a1268eca7fd9dcc60c0ba7e2057b54b8e483).

In this update, I decoupled the hashing functionality from ECDSA, allowing us to maintain previous compatibility. All existing tests are passing without requiring any function overloading.

Let me know what you think!

Best,
Morteza

**Comment by HLRichardson-Git** [comment]
Nice work, I like your approach!

I like that you wouldn't even need to make a change to `test_ecdsa.cpp` for this change. Additionally, if you'd like I would like to add your name/ git user name to the special thanks list in: `AUTHORS.md`. If you would like this would you let me know the name I should use?

Could you make a PR to this branch `release/0.6.2` as this is the last issue I had in mind for the 0.6.2 milestone? Then I will review everything again and make the push to main and change the version to 0.6.1 -> 0.6.2.

**Comment by mirzaim** [comment]
Hi Hunter,

Thanks so much! I’d be honored to be included in the special thanks list. You can use "Morteza Mirzaei" and my GitHub username "mirzaim" for that.

I’ll go ahead and make the PR to the release/0.6.2 branch.

**Comment by HLRichardson-Git** [comment]
I have merged the PR for this issue, like I noted in the PR I only had to add #include <functional> to gestalt/ecdsa.h for it to compile for me. 

Also I have added your name to `AUTHORS.md` as part of the special thanks section. 

Thanks again for taking on this issue, and implementing an elegant solution. I appreciate your effort. Hope to work with you more in the future on other issues for Gestalt!
